### ISSUE TITLE
Error - FindRecentlyPushedNewBranches, branch does not exist [repo_id: 64 name: main]

### ISSUE THREAD

**Issue by Flagelmann** [go-gitea/gitea#31163]
### Description

I'm getting a weird issue on several repositories.

The actual error is a **500 error**:

`FindRecentlyPushedNewBranches, branch does not exist [repo_id: 64 name: main]`

How to solve this? Should I modify the branch name directly on DB side? Is there any way to fix this from the Gitea web portal?

Actually, it seems that this new version has broken several repos at once. Until last week everything was good, clean and accessible. Now I'm getting this error on several repositories.

### Gitea Version

1.22.0

### Workaround

It has been fixed in 1.22-nightly (which will be 1.22.1 soon), please use the 1.22-nighly binary (https://dl.gitea.com/gitea/1.22-nightly/) or docker image.

**Comment by lunny** [comment]
Please try to run `sync branches` on admin panel.

**Comment by Flagelmann** [comment]
Yout mean the "Resynchronize pre-receive, update and post-receive hooks of all repositories."?

I have already tried multiple times it....nothing changed.

I also tried "Reinitialize all missing Git repositories for which records exist" under "Site Administration", nothing even with this.

I wish I will not have to rebuild the gitea instance from scratch.....so, re-import manually over 100 repositories (over 30 GB of data) and then also re-enable manually, by modifying manually each repository on database site the mirroring settings for each repo I have (as Gitea does not still have the way to backup these settings)....

**Comment by wxiaoguang** [comment]
> Please try to run `sync branches` on admin panel.

I do not think `FindRecentlyPushedNewBranches` logic is right. Anything wrong in that function shouldn't block end users seeing a complete page.

**Comment by lunny** [comment]
> Yout mean the "Resynchronize pre-receive, update and post-receive hooks of all repositories."?
> 
> I have already tried multiple times it....nothing changed.
> 
> I also tried "Reinitialize all missing Git repositories for which records exist" under "Site Administration", nothing even with this.
> 
> I wish I will not have to rebuild the gitea instance from scratch.....so, re-import manually over 100 repositories (over 30 GB of data) and then also re-enable manually, by modifying manually each repository on database side the mirroring settings for each repo I have (as Gitea does not still have the way to backup these settings)....

No. It's `Sync missed branches from git data to databases`

**Comment by Flagelmann** [comment]
Even with "Sync missed branches from git data to databases" nothing changed. It starts, but then nothing...

**Comment by wxiaoguang** [comment]
As a quick fix:  Ignore FindRecentlyPushedNewBranches err #31164 

If you could build your own instance, you could try to apply it to 1.22 branch.

Or when the backport is done and 1.22-nightly is ready, please try 1.22 nightly.

**Comment by Flagelmann** [comment]
I am using snapcraft to get updates of Gitea.

May I get the nightly build with snap or do I have just to wait the next build?

**Comment by yp05327** [comment]
As a quick fix it is ok, but I think the root reason is that sync branch has some problems.
The branch name is `main`, so it is default branch of repo 64?
Then what kind of this repo is? A fork, a mirror or a normal one?
Then why this branch can not be found in DB? During the creation of this repo, the main branch should has been insert into the DB.

**Comment by wxiaoguang** [comment]
Reopen, because the quick fix isn't complete. Need to figure out the root reason.

**Comment by wxiaoguang** [comment]
The root problem is that any branch could be renamed or deleted. Default branch may not exist in many cases.

So when running `FindRecentlyPushedNewBranches`, any non-existing branch should be ignored, but not reporting an error.

**Comment by yp05327** [comment]
I can only image one case that the last branch (it is also the default branch) was deleted, and there's no branch in the repo, this error would happen.
If the default branch is renamed, repo.DefaultBranch should also be renamed?
If there's two or more branches, then when the default branch was deleted, the other one will be the new default branch?

**Comment by Flagelmann** [comment]
Jfyi, I have fixed the issue by issuing a simple UPDATE query against the specific repository ID in the repository table, updating the default_branch value from "main" to "master". Just after doing that now I can access the affected repository.

I will have to check all of them one by one, as not all repository are using the "master" branch, some are using "main".

Here a sample query I used:

UPDATE repository SET default_branch = 'master' WHERE id = 100;

Finally, I dunno why this happened, as just before the upgrade to the latest build I was able to use the affected repositories without issues (and they were on master branch, not main, and I checked it in the commit history); after the upgrade several repositories have been "moved" to the main branch without a reason.

Wishing that this will not happen anymore in the next builds.

Regards

**Comment by lunny** [comment]
Maybe the algorithm of detection defulat branch when syncing branches into database are wrong. So that, all repositories with `master` as default branch but your default branch in app.ini is `main` will be wrong.

**Comment by Flagelmann** [comment]
Probably, for now I have reverted manually, as it was the only way to fix this without disruptions, each repository affected from the wrong default branch to the correct one (master, most of the times).

But now I'm a bit worried about next builds/updates. :D

**Comment by Flagelmann** [comment]
No, actually there was even another repository that was not on master (still figuring out which was the previous one) but it was put on main also.

**Comment by spiritLHLS** [comment]
> Jfyi, I have fixed the issue by issuing a simple UPDATE query against the specific repository ID in the repository table, updating the default_branch value from "main" to "master". Just after doing that now I can access the affected repository.
> 
> I will have to check all of them one by one, as not all repository are using the "master" branch, some are using "main".
> 
> Here a sample query I used:
> 
> UPDATE repository SET default_branch = 'master' WHERE id = 100;
> 
> Finally, I dunno why this happened, as just before the upgrade to the latest build I was able to use the affected repositories without issues (and they were on master branch, not main, and I checked it in the commit history); after the upgrade several repositories have been "moved" to the main branch without a reason.
> 
> Wishing that this will not happen anymore in the next builds.
> 
> Regards

I'm having the same problem, hopefully the next new release fixes this and does it automatically (since my gitea is mainly used for backups, there are too many affected repositories for me to have the time to detect and manually fix them all).

**Comment by Flagelmann** [comment]
Well, I checked the 180+ repositories I have right now on my Gitea instance and I would say that 1/3 of them have been affected by this issue. Now I have fixed them, but obviously I will monitor after next update if this will happen again.

**Comment by eeyrjmr** [comment]
This is what I faced as well, luckily only one ...

**Comment by Flagelmann** [comment]
> This is what I faced as well, luckily only one ...

Lucky one :D

**Comment by eeyrjmr** [comment]
I wonder if it is all the same issue (from gitea perspective).
Basically gitea was unable to find the default branch (master or main) and hence a 500 page. If you append /issue to the URL you get to the other webUI of gitea and from there you can view branches etc. for me it was not showing master (only the other feature branches )

A git clone works and git remote -r would list the feature branches but I also had what appeared as two origins and git would provide a WARNING (while this was an ERROR for gitea).

If you Google around you will find this can occur with git and it tolerates it but gitea needs to be a bit more deterministic especially with the db sinking 

My workaround was to create a new branch from the cli from what git considered master, remove the master branches (local and origin), push to origin, from gitea /issue page then change the default branch to the new branch and everything works. Finally rename this tmp branch to master


Obviously this is a tedious workaround to get the git repo in alignment so the question then is.. what should gitea do and what is causing these shadow branches

**Comment by Flagelmann** [comment]
I did not know about the /issue URL part. Good to know.

Probably this is just a Gitea-side issue as I checked, in my case, and the git repos were still there untouched, but not visible.

Actually in my case it was much quicker to identify which was the wrong value on DB side and update it directly for all the affected repo IDs.

**Comment by Cat7373** [comment]
I also encountered this problem, after upgrading to 1.22.0

**Comment by wxiaoguang** [comment]
> I also encountered this problem, after upgrading to 1.22.0

Please try 1.22-nightly , and/or try to "Sync branches" from the admin dashboard panel

**Comment by yp05327** [comment]
Maybe this is caused by auto rename branch name
After the rename, the branch name in DB is not changed
![image](https://github.com/go-gitea/gitea/assets/18380374/c2b54396-aa8d-4ca5-8f1c-b3aa7e7f4f36)

**Comment by yp05327** [comment]
Does this related to #28056 ?
User didn't notice this issue, so branches are not synced since updated to v1.21, and errors occurred after upgrade to v1.22?

**Comment by Flagelmann** [comment]
Actually in my case it was not due to an auto-renaming of the branches, as I had, suddenly after the auto-upgrade (using snap image) I had 40-50 repositories not usable anymore using Gitea, then I fixed each of them directly in DB and after the value updated on DB side I did just an F5 on the page of the affected repo and it turned back working as expected.

**Comment by theAkito** [comment]
100% agree with OP. Same experience here.

**Comment by wxiaoguang** [comment]
> 100% agree with OP. Same experience here.

It has been fixed in 1.22-nightly (which will be 1.22.1 soon)

**Comment by theAkito** [comment]
> > 100% agree with OP. Same experience here.
> 
> It has been fixed in 1.22-nightly (which will be 1.22.1 soon), please use the 1.22-nighly binary (https://dl.gitea.com/gitea/1.22-nightly/) or docker image.
> 

Ah, I see, I thought it would be backported into the 1.22 branch.

Is this Nightly release production ready or would it be rather advisable to wait for 1.22.1?

Is there a release date for 1.22.1?

**Comment by wxiaoguang** [comment]
> Ah, I see, I thought it would be backported into the 1.22 branch.

Yes, it has been backported to 1.22, so 1.22-nightly contains it.

> Is this Nightly release production ready or would it be rather advisable to wait for 1.22.1?

Yes, because 1.22-nighty will be 1.22.1 soon. It is exactly for production so no need to wait.

> Is there a release date for 1.22.1?

Maybe in a several days.

**Comment by theAkito** [comment]
@wxiaoguang

Which branch ot commit do I need to check out from Git to get exactly the state of `1.22-nightly`? I want to build my own Docker image.

**Comment by delvh** [comment]
`release/1.22`

**Comment by delvh** [comment]
(± one day as `nightly` is probably built only once a day as the name implies)

**Comment by theAkito** [comment]
Nice, seems to work now and the issue is worked around. Thank you!

**Comment by delvh** [comment]
The issue seems to be fixed.

**Comment by yp05327** [comment]
The root reason is not fixed.

**Comment by theAkito** [comment]
@yp05327 I agree.

I even specifically said

> the issue is worked around

because it is *not* fixed. 😄

**Comment by wxiaoguang** [comment]
I believe the root problem needs a new issue to clarify the problem, but not keeping discussing "FindRecentlyPushedNewBranches" here .....

The FindRecentlyPushedNewBranches is indeed fixed, it sounds like a workaround but the approach is right: to make the system robust. No matter there is the root problem or not, the fix is still right and necessary.

**Comment by lunny** [comment]
Let's open a new issue.

**Comment by theAkito** [comment]
> Let's open a new issue.

Which is it? I don't see a related new issue in the list.

I'll create a new one, then.

[Here it is](https://github.com/go-gitea/gitea/issues/31471).

**Comment by amirad1** [comment]
It seems in the 1.22.1 version, the 500 error is fixed but I still get the error log ? should I ignore it or not ?

**Comment by lunny** [comment]
> It seems in the 1.22.1 version, the 500 error is fixed but I still get the error log ? should I ignore it or not ?

Please open a new issue and paste your error logs.

**Comment by wxiaoguang** [comment]
@lunny you have asked so above  https://github.com/go-gitea/gitea/issues/31163#issuecomment-2185600813 and there is already one https://github.com/go-gitea/gitea/issues/31163#issuecomment-2185951268 : #31471
